interp/src/interpreter/interpret_group.rs

```diff
@@ -1,357 +1,479 @@
 //! Used for the command line interface.
 //! Only interprets a given group in a given component
 
-use crate::{environment::Environment, primitives, update::UpdateQueue};
+use crate::environment::Environment;
+
+use crate::utils::{AssignmentRef, CellRef, OutputValueRef};
+use crate::values::{OutputValue, ReadableValue, TimeLockedValue, Value};
 use calyx::{
-    errors::{Error, FutilResult},
-    ir,
-    ir::CloneName,
+    errors::FutilResult,
+    ir::{self, RRC},
 };
-use std::collections::HashMap;
-// use std::rc::Rc;
-
-/// Stores information about the component and group to interpret.
-/// Might be better to make this a subset of a trait implemented by all interpreters, later on
-
-/// Get the name of the component to interpret from the context.
-fn _get_component(
-    ctx: ir::Context,
-    component: &str,
-) -> FutilResult<ir::Component> {
-    match ctx.components.into_iter().find(|c| c.name.id == *component) {
-        Some(comp) => Ok(comp),
-        None => Err(Error::Undefined(
-            ir::Id::from(component.to_string()),
-            "component".to_string(),
-        )),
-    }
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+use std::iter;
+use std::rc::Rc;
+
+#[allow(unused_imports)]
+use crate::primitives::{
+    Execute, ExecuteBinary, ExecuteStateful, ExecuteUnary,
+};
+#[derive(Debug, Clone, Default)]
+struct DependencyMap<'a> {
+    map: HashMap<*const ir::Port, HashSet<AssignmentRef<'a>>>,
 }
 
-/// Construct a map from cell ids to a map from the cell's ports' ids to the ports' values
-fn _construct_map(
-    cells: &[ir::RRC<ir::Cell>],
-) -> HashMap<ir::Id, HashMap<ir::Id, u64>> {
-    let mut map = HashMap::new();
-    for cell in cells {
-        let cb = cell.borrow();
-        let mut ports: HashMap<ir::Id, u64> = HashMap::new();
-
-        match &cb.prototype {
-            // A Calyx constant cell's out port is that constant's value
-            ir::CellType::Constant { val, .. } => {
-                ports.insert(ir::Id::from("out"), *val);
-                map.insert(cb.clone_name(), ports);
+impl<'a> DependencyMap<'a> {
+    fn from_assignments<I: Iterator<Item = &'a ir::Assignment>>(
```

Comment:
Why not implement this using the `From` trait? Do the lifetimes make it complicated?

Reply:
Good point! I'm not sure why I didn't make it that originally since it works fine